cloudinit/sources/DataSourceVMware.py

```diff
@@ -292,6 +344,543 @@ def redact_keys(self):
+        """
+        check to see if there is data via env vars
+        """
+        md, ud, vd = None, None, None
+        if os.environ.get(VMX_GUESTINFO, ""):
+            md_in_json_or_yaml = guestinfo_envvar("metadata")
+            ud = guestinfo_envvar("userdata")
+            vd = guestinfo_envvar("vendordata")
+
+            if md_in_json_or_yaml:
+                md = load_json_or_yaml(md_in_json_or_yaml)
+        return (md, ud, vd)
+
+    def get_guestinfo_data_fn(self):
+        """
+        check to see if there is data via the guestinfo transport
+        """
+        md, ud, vd = None, None, None
+        if self.vmware_rpctool:
+            md_in_json_or_yaml = guestinfo("metadata", self.vmware_rpctool)
+            ud = guestinfo("userdata", self.vmware_rpctool)
+            vd = guestinfo("vendordata", self.vmware_rpctool)
+
+            if md_in_json_or_yaml:
+                md = load_json_or_yaml(md_in_json_or_yaml)
+        return (md, ud, vd)
+
+    def get_vmware_imc_data_fn(self):
+        """
+        check to see if there is data via vmware guest customization
+        """
+        md, ud, vd = None, None, None
+
+        # Check if vmware guest customization is enabled.
+        allow_vmware_cust = self.is_vmware_cust_enabled()
+        allow_raw_data_cust = self.is_raw_data_cust_enabled()
+        if not allow_vmware_cust and not allow_raw_data_cust:
+            LOG.debug("Customization for VMware platform is disabled")
+            return (md, ud, vd)
+
+        # Check if "VMware Tools" plugin is available.
+        if not is_cust_plugin_available():
+            return (md, ud, vd)
+
+        # Wait for vmware guest customization configuration file.
+        cust_cfg_file = self.get_cust_cfg_file()
+        if cust_cfg_file is None:
+            return (md, ud, vd)
+
+        # Check what type of guest customization is this.
+        cust_cfg_dir = os.path.dirname(cust_cfg_file)
+        cust_cfg = parse_cust_cfg(cust_cfg_file)
+        (is_vmware_cust_cfg, is_raw_data_cust_cfg) = get_cust_cfg_type(
+            cust_cfg
+        )
+
+        # Get data only if guest customization type and flag matches.
+        if is_vmware_cust_cfg and allow_vmware_cust:
+            LOG.debug("Getting data via VMware customization configuration")
+            (md, ud, vd) = self.get_data_from_vmware_cust_cfg(
+                cust_cfg, cust_cfg_dir
+            )
+        elif is_raw_data_cust_cfg and allow_raw_data_cust:
+            LOG.debug(
+                "Getting data via VMware raw cloudinit data "
+                "customization configuration"
+            )
+            (md, ud, vd) = self.get_data_from_raw_data_cust_cfg(cust_cfg)
+        else:
+            LOG.debug("No allowed customization configuration data found")
+
+        # Clean customization configuration file and directory
+        util.del_dir(cust_cfg_dir)
+        return (md, ud, vd)
+
+    def is_vmware_cust_enabled(self):
```

Comment:
Please document this function. It may appear obvious, but for newcomers, they may not know this is specific to IMC. In fact, it would be good for all the sub-transport functions to have some prefix that indicates what they are, ex.

```python
def imc_is_cust_enabled(self):
```

Reply:
All the imc specified functions have been moved to helpers.vmware.imc.guestcust_util.py, thus when calling them with a guestcust_util. prefix in DataSourceVMware.py, we know they are imc specified functions.